c++/src/capnp/serialize-async.c++

```diff
@@ -252,39 +252,77 @@ struct WriteArrays {
 
-template <typename WriteFunc>
-kj::Promise<void> writeMessageImpl(kj::ArrayPtr<const kj::ArrayPtr<const word>> segments,
-                                   WriteFunc&& writeFunc) {
+// Helper function that allocates and fills the pointed-to table with info about the segments and
+// populates the pieces array with pointers to the segments.
+void fillWriteArraysWithMessage(kj::ArrayPtr<const kj::ArrayPtr<const word>> segments,
+                                kj::Array<_::WireValue<uint32_t>>* table,
+                                kj::ArrayPtr<kj::ArrayPtr<const byte>> pieces) {
   KJ_REQUIRE(segments.size() > 0, "Tried to serialize uninitialized message.");
 
-  WriteArrays arrays;
-  arrays.table = kj::heapArray<_::WireValue<uint32_t>>((segments.size() + 2) & ~size_t(1));
+  *table = kj::heapArray<_::WireValue<uint32_t>>((segments.size() + 2) & ~size_t(1));
 
   // We write the segment count - 1 because this makes the first word zero for single-segment
   // messages, improving compression.  We don't bother doing this with segment sizes because
   // one-word segments are rare anyway.
-  arrays.table[0].set(segments.size() - 1);
+  (*table)[0].set(segments.size() - 1);
   for (uint i = 0; i < segments.size(); i++) {
-    arrays.table[i + 1].set(segments[i].size());
+    (*table)[i + 1].set(segments[i].size());
   }
   if (segments.size() % 2 == 0) {
     // Set padding byte.
-    arrays.table[segments.size() + 1].set(0);
+    (*table)[segments.size() + 1].set(0);
   }
 
-  arrays.pieces = kj::heapArray<kj::ArrayPtr<const byte>>(segments.size() + 1);
-  arrays.pieces[0] = arrays.table.asBytes();
-
+  KJ_ASSERT(pieces.size() == segments.size() + 1, "incorrectly sized pieces array during write");
+  pieces[0] = table->asBytes();
   for (uint i = 0; i < segments.size(); i++) {
-    arrays.pieces[i + 1] = segments[i].asBytes();
+    pieces[i + 1] = segments[i].asBytes();
   }
+}
+
+template <typename WriteFunc>
+kj::Promise<void> writeMessageImpl(kj::ArrayPtr<const kj::ArrayPtr<const word>> segments,
+                                   WriteFunc&& writeFunc) {
+  KJ_REQUIRE(segments.size() > 0, "Tried to serialize uninitialized message.");
+
+  WriteArrays arrays;
+  arrays.pieces = kj::heapArray<kj::ArrayPtr<const byte>>(segments.size() + 1);
+  fillWriteArraysWithMessage(segments, &arrays.table, arrays.pieces);
 
   auto promise = writeFunc(arrays.pieces);
 
   // Make sure the arrays aren't freed until the write completes.
   return promise.then(kj::mvCapture(arrays, [](WriteArrays&&) {}));
 }
 
+template <typename WriteFunc>
+kj::Promise<void> writeMessagesImpl(
+    kj::ArrayPtr<kj::ArrayPtr<const kj::ArrayPtr<const word>>> messages, WriteFunc&& writeFunc) {
+  KJ_REQUIRE(messages.size() > 0, "Tried to serialize zero messages.");
+
+  // Construct separate tables for each message, but one shared array of pieces that all the
+  // messages get combined into.
+  auto tables = kj::heapArray<kj::Array<_::WireValue<uint32_t>>>(messages.size());
```

Comment:
I suppose technically it would be possible to allocate one big array containing all the tables, then chop it up... but I guess that'd make the code considerably more complicated for not an enormous benefit.

Reply:
Eh, it does make the code a little tougher to read, but it's not too bad. I'll throw that in.